File: c/rational_numbers.c
#include "rational_numbers.h"

#include <math.h>
#include <stdlib.h>

int gcd(int a, int b);

rational_t add(rational_t a, rational_t b) {
  rational_t r;
  r.numerator = a.numerator * b.denominator + b.numerator * a.denominator;
  r.denominator = a.denominator * b.denominator;
  return reduce(r);
}

rational_t subtract(rational_t a, rational_t b) {
  rational_t r;
  r.numerator = a.numerator * b.denominator - b.numerator * a.denominator;
  r.denominator = a.denominator * b.denominator;
  return reduce(r);
}

rational_t multiply(rational_t a, rational_t b) {
  rational_t r;
  r.numerator = a.numerator * b.numerator;
  r.denominator = a.denominator * b.denominator;
  return reduce(r);
}

rational_t divide(rational_t a, rational_t b) {
  rational_t r;
  r.numerator = a.numerator * b.denominator;
  r.denominator = a.denominator * b.numerator;
  return reduce(r);
}

rational_t absolute(rational_t r) {
  return reduce((rational_t){.denominator = abs(r.denominator),
                             .numerator = abs(r.numerator)});
}

rational_t exp_rational(rational_t a, int16_t n) {
  int d = n < 0 ? a.numerator : a.denominator;
  int nu = n < 0 ? a.denominator : a.numerator;
  n = abs(n);
  rational_t r =
      (rational_t){.denominator = pow(d, n), .numerator = pow(nu, n)};
  return reduce(r);
}

float exp_real(uint16_t x, rational_t r) {
  return pow(pow(x, r.numerator), 1. / r.denominator);
}

rational_t reduce(rational_t r) {
  if (r.numerator == 0) {
    return (rational_t){.denominator = 1, .numerator = 0};
  }
  int d = abs(r.denominator);
  int n = abs(r.numerator);
  int g = gcd(d, n);
  int sign = r.denominator * r.numerator < 0 ? -1 : 1;
  return (rational_t){.denominator = d / g, .numerator = n * sign / g};
}

int gcd(int a, int b) {
  if (a % b == 0)
    return b;

  return gcd(b, a % b);
}

File: lua/rational-numbers.lua
local function gcd(a, b)
    if a % b == 0 then
        return b
    else
        return gcd(b, a % b)
    end
end

local function reduce(a)
    local g = gcd(a[1], a[2])
    return { a[1] // g, a[2] // g }
end

local function add(a, b)
    return reduce({ a[1] * b[2] + a[2] * b[1], a[2] * b[2] })
end

local function subtract(a, b)
    b[1] = b[1] * -1
    return add(a, b)
end

local function multiply(a, b)
    return reduce({ a[1] * b[1], a[2] * b[2] })
end

local function divide(a, b)
    return reduce({ a[1] * b[2], a[2] * b[1] })
end

local function abs(a)
    a[1] = math.abs(a[1])
    a[2] = math.abs(a[2])
    return a
end

local function exp_rational(a, p)
    return reduce({ a[1] ^ p, a[2] ^ p })
end

local function exp_real(p, a)
    return (p ^ a[1]) ^ (1 / a[2])
end

return {
    add = add,
    subtract = subtract,
    multiply = multiply,
    divide = divide,
    abs = abs,
    exp_rational = exp_rational,
    exp_real = exp_real,
    reduce = reduce,
}
